Interpret binops

brson opened the pull request:
This is wip for https://github.com/dada-lang/dada/issues/8. It is mergeable as is but there are a number of obvious things left do do, such as:

- write negative tests for bogus type / op combos
- do something on divide by zero. what?
- put todo!s on ops that I know should be implemented that I don't intend to think about yet, like float eq

There are also some seemingly missing ops like negation, not-equals, less-than-or-equals, greater-than-or-equals. And I can't find a way to create signed ints, so no tests for them.

I'm happy to restructure the code any way you like.

brson commented:
I wonder how thorough you want the test suite to be. I've now added 5 distinct tests each for a single operator that doesn't work with bools. There's lots more I can add like that. Maybe you don't want to deal with so many minor test cases yet.

nikomatsakis commented on dada_tests/interpret/ops_uint.dada:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    # wrapping overflow
+    if uint_max + 2 == 1 {
```
we should decide what to do here =) bah, the annoying things. Maybe return an Err for now?

nikomatsakis reviewed (approved):
This looks good to me, I'd be up to land it as is, though we should open a FIXME to decide what to do about overflow.

nikomatsakis commented:
> do something on divide by zero. what?

I think overflow and divide by zero should yield an error for now (i.e., return `Err` for the `eyre::Result<Value>`)

nikomatsakis commented:
> I wonder how thorough you want the test suite to be. I've now added 5 distinct tests each for a single operator that doesn't work with bools. There's lots more I can add like that. Maybe you don't want to deal with so many minor test cases yet.

No strong opinion, I guess. =) I am happy to have more tests.

nikomatsakis commented:
> And I can't find a way to create signed ints, so no tests for them.

Yeah I was too lazy to support these. =) I guess we should add some kind of "suffixes" to the parser (we shouldn't need new tokens for that, we can use the "skipped whitespace" flag) for the time being.

brson commented:
@nikomatsakis I've changed oveflow to errors, divide by zero to errors, added a special case error for signed division overflow and added negative tests for bad string and unit binops.

Other types like floats are left for the future. I'll file a few issues.

nikomatsakis commented:
bors r+

Cool =)
